### PR TITLE
Seabbs/issue199

### LINKED ISSUE #199: Support for {touchstone} based benchmarking.
`{touchstone}` is an R package for benchmarking functionality in a branch against the PR target. We make use of it in [`{EpiNow2}`](https://github.com/epiforecasts/EpiNow2/tree/main/touchstone)  (mainly implemented by @sbfnk) were it allows us to make sure we don't accidently introduce performance regressions. Its already proved useful in identifying [issues with new features](https://github.com/epiforecasts/EpiNow2/pull/363#issuecomment-1428427959). 

It would be great to have in `{epinowcast}`. In the first instance wrapping it around the examples in `inst/examples` seems like it could be a good idea?

### PR DESCRIPTION
**Pull request by seabbs**
This PR closes #199 by adding `{touchstone}` benchmarks using the package script examples. See #238 for a test implementation.

### CODE REVIEW

File: test.R [hunk 1]
```diff
@@ -0,0 +1,7 @@
+library(epinowcast)
+
+ nowcast <- enw_example("nowcast")
+ summarised_nowcast <- summary(nowcast)
+ obs <- enw_example("observations")
+
+enw_score_nowcast(summarised_nowcast, obs)
```
Comment (adrian-lison) [inline test.R:7]: What is this script for?

Comment (seabbs) [reply test.R:7]: ack - that is left over from making a reprex for #198 will remove.

**Comment by adrian-lison** [comment]
> The reason this is failing at the moment is because I believe it has to have a version on the target branch in order to work (so here`develop`.

So you think it will work after we have merged this in develop? If we are scared, we could also create another test branch from this one here, make some changes and test a pull request from the test branch into this one.

File: test.R (hunk 1 above)
Comment (seabbs) [reply test.R]: Removed in https://github.com/epinowcast/epinowcast/pull/200/commits/f2d6fc03895d1448749f9fec2f7a2ca4c818df35

**Comment by seabbs** [comment]
As suggested by @adrian-lison this is now being tested in #203

**Comment by seabbs** [comment]
Proving to be a bit of a fiddle to get right.

**Comment by adrian-lison** [comment]
@seabbs I looked into this now and it seems that the current point of failure is really the function `pin_assets`. It works neither when just using `"touchstone"` or just using `"inst/examples"`, and the failure happens somewhere in the call to `pin_assets` (see [here](https://github.com/lorenzwalthert/touchstone/blob/e4b325587c6425f00fbf6dd989132bd50532a9e2/R/utils.R)), which then triggers the following error from the `cli` package ([see here](https://rdrr.io/github/r-pkgs/boxes/src/R/ansi-hyperlink.R)):
```
Error in vapply(X, FUN, FUN.VALUE = character(1), ..., USE.NAMES = USE.NAMES) : 
  values must be type 'character',
 but FUN(X[[1]]) result is type 'list'
Calls: <Anonymous> ... transform -> make_link -> make_link_file -> vcapply -> vapply
```

I haven't figured out yet which call exactly it is (maybe more deep down the hierarchy, because all the direct calls to `cli` in the function seem alright) and what is causing this.

One general guess though I had might be versioning of `{touchstone}`. At the moment, we have this in the workflow
`- uses: lorenzwalthert/touchstone/actions/receive@main`
but maybe we have to specify the correct version or use the v1 wildcard: https://github.com/lorenzwalthert/touchstone/blob/main/actions/README.md ?

**Comment by adrian-lison** [comment]
My best guess is that one or several of the directories that we want to pin cannot be found and then it is doing something really weird when trying to write a warning on this using `cli`, see [here](https://github.com/lorenzwalthert/touchstone/blob/e4b325587c6425f00fbf6dd989132bd50532a9e2/R/utils.R#L213)
```
if (!all(valid_dirs)) {
    cli::cli_warn(paste0(
      "The following asset{?s} could not be found and will ",
      "not be copied: {.path {dirs[!valid_dirs]}}"
    ))
  }
  ```
  
  somehow, if trying to format a path which is not a character but a list, then we get exactly the error we have. But I have no idea how this could happen, I cannot reproduce at this function level.

**Comment by adrian-lison** [comment]
@seabbs It must really be a version issue though. Did you know that `{touchstone}` is recently also failing with the same error in `EpiNow2`? https://github.com/epiforecasts/EpiNow2/actions/runs/4366678150/jobs/7637021308?pr=372

So maybe we can go back to an older version of touchstone? And we should potentially inform the developer.

**Comment by adrian-lison** [comment]
Tried a workaround that avoides `pin_assets` by sourcing the relevant files directly (works as long they agree in all branches), but now we get new errors. Currently I am stuck with [this one](https://github.com/epinowcast/epinowcast/actions/runs/4501160645/jobs/7921222512?pr=203). Very weird.

**Comment by seabbs** [comment]
> but maybe we have to specify the correct version or use the v1 wildcard:

Really good point about using standard versions. This seems like it might well be it (especially due to the change in `EpiNow2` from working -> not. 

> somehow, if trying to format a path which is not a character but a list, then we get exactly the error we have. But I have no idea how this could happen, I cannot reproduce at this function level.

Great job digging and yeah I have no idea what is happening here. 

By the way for loca testing I have been finding `act` really helpful for GitHub Actions: https://github.com/nektos/act

**Comment by seabbs** [comment]
> ut now we get new errors. Currently I am stuck with [this one](https://github.com/epinowcast/epinowcast/actions/runs/4501160645/jobs/7921222512?pr=203). Very weird.

This is very weird

**Comment by seabbs** [comment]
> but maybe we have to specify the correct version or use the v1 wildcard:

It looks like the last release of this was in Jan 2022 but a bug was fixed due to later releases of `cli` in November 2022. This could maybe be an issue. I have reverted to use `main` for both actions.

**Comment by seabbs** [comment]
So by comparing with working version @sbfnk implemented for `EpiNow2` I have got this working. The core of the issue appears to have been trying to source within the benchmark. Moving to just a function call appears to have got things working (🤞🏻)

**Comment by seabbs** [comment]
All finally working now so just a matter of fine-tuning what we want to run in the benchmarks. Annoyingly in needs to recompile the model each time and so this slows everything down. It doesn't seem like we can avoid this either.

**Review by seabbs — COMMENTED** [review]
This all appears to be working and I think the current benchmarks are a good MVP. Planning on merging as a self-review and we can iterate.

**Pull request merged**